segmentation/include/pcl/segmentation/extract_clusters.h

```diff
@@ -112,69 +119,18 @@ namespace pcl
       PCL_ERROR ("[pcl::extractEuclideanClusters] Number of points in the input point cloud (%lu) different than normals (%lu)!\n", cloud.points.size (), normals.points.size ());
       return;
     }
-
-    // Create a bool vector of processed point indices, and initialize it to false
-    std::vector<bool> processed (cloud.points.size (), false);
-
-    std::vector<int> nn_indices;
-    std::vector<float> nn_distances;
-    // Process all points in the indices vector
-    for (std::size_t i = 0; i < cloud.points.size (); ++i)
-    {
-      if (processed[i])
-        continue;
-
-      std::vector<unsigned int> seed_queue;
-      int sq_idx = 0;
-      seed_queue.push_back (static_cast<int> (i));
-
-      processed[i] = true;
-
-      while (sq_idx < static_cast<int> (seed_queue.size ()))
-      {
-        // Search for sq_idx
-        if (!tree->radiusSearch (seed_queue[sq_idx], tolerance, nn_indices, nn_distances))
-        {
-          sq_idx++;
-          continue;
-        }
-
-        for (std::size_t j = 1; j < nn_indices.size (); ++j)             // nn_indices[0] should be sq_idx
-        {
-          if (processed[nn_indices[j]])                         // Has this point been processed before ?
-            continue;
-
-          //processed[nn_indices[j]] = true;
-          // [-1;1]
-          double dot_p = normals.points[i].normal[0] * normals.points[nn_indices[j]].normal[0] +
-                         normals.points[i].normal[1] * normals.points[nn_indices[j]].normal[1] +
-                         normals.points[i].normal[2] * normals.points[nn_indices[j]].normal[2];
-          if ( std::acos (std::abs (dot_p)) < eps_angle )
-          {
-            processed[nn_indices[j]] = true;
-            seed_queue.push_back (nn_indices[j]);
-          }
-        }
-
-        sq_idx++;
-      }
-
-      // If this queue is satisfactory, add to the clusters
-      if (seed_queue.size () >= min_pts_per_cluster && seed_queue.size () <= max_pts_per_cluster)
-      {
-        pcl::PointIndices r;
-        r.indices.resize (seed_queue.size ());
-        for (std::size_t j = 0; j < seed_queue.size (); ++j)
-          r.indices[j] = seed_queue[j];
-
-        // These two lines should not be needed: (can anyone confirm?) -FF
-        std::sort (r.indices.begin (), r.indices.end ());
-        r.indices.erase (std::unique (r.indices.begin (), r.indices.end ()), r.indices.end ());
-
-        r.header = cloud.header;
-        clusters.push_back (r);   // We could avoid a copy by working directly in the vector
-      }
-    }
+    auto lambda = [&](int &i, int &j, std::vector<int>& nn_indices) -> bool {
+      //processed[nn_indices[j]] = true;
+      // [-1;1]
+      double dot_p = normals.points[i].normal[0] * normals.points[nn_indices[j]].normal[0] +
```

Comment:
This would be a good place for executor usage, we could even do dot product using omp (though it would be only benificial if the cloud size is quite large).

I think we might encounter this problem multiple places when providing SIMD alternatives. We should just give a warning to the user and set the last element to 0. 
Leaving it to the user whether he wants performance or preserve the last element.
We can also consider preserving the last element of the user is fine with the slight performance penalty.

Reply:
> I think we might encounter this problem multiple places when providing SIMD alternatives. We should just give a warning to the user and set the last element to 0.

We don't provide explicit access to the 4th component of the normals, suggesting it should not be touched by users. I would set it to zero with no warning.